registry/client/repository.go

```diff
@@ -632,6 +635,50 @@ func (ms *manifests) Delete(ctx context.Context, dgst digest.Digest) error {
 	return HandleErrorResponse(resp)
 }
 
+func (ms *manifests) digestFromResponse(ctx context.Context, resp *http.Response, manifestURL string) (string, error) {
+	// The specs are ambivalent on Docker-Content-Digest header: on the one hand, the docs at https://docs.docker.com/registry/spec/api/#digest-header
+	// mention that "any response may include a Docker-Content-Digest header"  and on the other hand, the docs at
+	// https://docs.docker.com/registry/spec/api/#put-manifest say that a PUT request should return a Docker-Content-Digest header.
```

Comment:
Note that the docs at https://docs.docker.com/registry/spec/api/ are built from the [files in this repository](https://github.com/distribution/distribution/blob/597e8a6b3459da8747bad6fefa880e9c60a7d762/docs/spec/api.md), and describe the "Docker Registry HTTP API V2" specification, which has been superseded by the [OCI distribution spec](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md).

From the OCI distribution spec for the [`PUT /v2/<name>/manifests/<reference>` endpoint](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#pushing-manifests)

> `<name>` is the namespace of the repository, and the `<reference>` MUST be either a) a digest or b) a tag.

and

> The `Docker-Content-Digest` header returns the canonical digest of the uploaded blob, and MUST be equal to the client provided digest. Clients MAY ignore the value but if it is used, the client SHOULD verify the value against the uploaded blob data.

So I guess there's 2 possible situations here;

- when pushing the manifest with a digest as `<reference>`, the header MUST be the same as that digest
- when pushing the manifest with a _name_ as `<reference>`, the header must contain the canonical digest of the data

So, I'm not sure if it's possible to validate anything on the client side if the registry does not return a `Docker-Content-Digest` header (at least not with the current spec), other than the `201` status code, and trusting that the data was sent correctly;

> The registry MUST store the manifest in the exact byte representation provided by the client. Upon a successful upload, the registry MUST return response code 201 Created,

With the current spec describing these headers as [OPTIONAL](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#legacy-docker-support-http-headers);

> Because of the origins this specification, the client MAY encounter Docker-specific headers, such as `Docker-Content-Digest`, or `Docker-Distribution-API-Version`. These headers are OPTIONAL and clients SHOULD NOT depend on them.

So the only correct option is to _ignore_ missing `Docker-Content-Digest` headers, and to skip validation.

That said, it feels like an omission in the OCI spec that there's no option to validate, so perhaps this should be raised as (per this discussion) it looks like there's no real alternative to verify if the registry calculated the expected digest.

Perhaps this endpoint needs a `?digest=<digest>` query-parameter, as is [provided for `PUT /v2/<name>/blobs/uploads/<reference>?digest=<digest>`](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc3/spec.md#endpoints), which would allow the registry to reject the request if the calculated ("canonical") digest does not match the digest that's provided by the client.

Reply:
I agree with your analysis that the absence of a digest should not result in an error - it should be up to the client to determine if a digest is expected.